decisions/DR012-Entity-reference-validation-workflow.md

```diff
@@ -0,0 +1,355 @@
+- Clear language bindings based on existing patterns.
+- No way to construct an invalid entity reference.
+
+#### Cons
+
+- Host must use exceptions to test validity, increasing boilerplate
+  and reducing performance.
+
+### Option 3
+
+`isEntityReference` hidden from the host. `createEntityReferenceIfValid`
+is instead used to return an `optional<EntityRef>` in C++
+(`None`/`EntityRef` in Python; `NULL` (and error code)/`EntityRef_h` in
+C).
+
+#### Pros
+
+- Only one way to accomplish validation and construction.
+- Validation and construction are performed in one operation.
+- Lowest host boilerplate for many workflows, utilising modern C++.
+- If using  `optional::value()` to extract the `EntityRef` before
+  passing to API methods, then safety is guaranteed, since a
+  `bad_optional_access` will be thrown if invalid (though less
+  informative than a custom exception).
+- Pybind has built-in support for `std::optional`, translating "not set"
+  to `None`.
+
+#### Cons
+
+- Very slightly less performant than `isEntityReference` in the invalid
+  reference case (at the risk of second-guessing compiler
+  optimisations).
+- Not immediately obvious in Python that the host must test the returned
+  value is valid (not `None`).
+- Uses a different optional API to `TraitsData`, decided
+  in [DR010](DR010-TraitsData-value-retrieval-API.md).
+  Though many of the arguments do not hold for this case (e.g.
+  direct-assignment to existing variables; detailed result codes).
+- Using the dereferencing operator (`*`) on a `std::optional` is unsafe
+  and loses the compile-time guarantee that the entity reference is
+  safe before being passed to API methods.
+- Language bindings to C and Python have a different interface than C++.
+
+## Outcome
+
+Based on the sketches in the Appendix (below) and the various pros and
+cons we have decided on supporting all workflows, i.e. implementing
+both Option 1/2 and 3.
+
+The number of potential host workflows, which we have little visibility
+on at present, mean that all options have enough merits to warrant
+first-class inclusion in the core API.
+
+## Appendix
+
+### Sketch - Publishing from a FileWidget
+
+We assume four cases
+
+* The widget stores an entity reference, and it can be published
+  in-process.
+* The widget stores an entity reference, but it must be published
+  out-of-process.
+* The widget does not store an entity reference, and it can be published
+  in-process.
+* The widget does not store an entity reference, and it must be
+  published out-of-process.
+
+Specifically, the following workflow assumes that, if the publish
+involves an expensive render operation, then the host application will
+spawn a subprocess (possibly at some later time and/or on a different
+machine) to perform the operation. In this case, we must pass the
+original string to the subprocess to be re-validated when it spawns.
+
+#### Option 1
+
+Here we use `isEntityReference` to check if the `file_widget` should
```

Comment:
[consider] Just wondering if the 'expensive process' bit might be a bit of a distraction from the main topic of "I dont need the ref yet". What about something like:

```python
path_only = user_input.get_value()

if manager.isEntityReference(path_only):
    raise ValueError("Sorry, entity references are not supported here")

export_to_file(path_only)
```

Reply:
I split the example in two in ab34d19f, since as you say there is the "I don't want an entity reference (yet)", as well as the "I want to take a different code path depending if it's an entity reference"